libs/stats/odc/stats/tasks.py

```diff
@@ -142,39 +206,30 @@ def msg_default(msg):
         if isinstance(temporal_range, str):
             temporal_range = DateTimeRange(temporal_range)
 
-        cfg: Dict[str, Any] = dict(
-            grid=self._grid, freq=self._frequency,
-        )
-
-        query = dict(product=product)
+        if dss is None:
+            dss, n_dss, cfg  = self._get_dss(dc, product, msg, temporal_range, tiles)
+        else:
 
-        if tiles is not None:
-            (x0, x1), (y0, y1) = tiles
-            msg(f"Limit search to tiles: x:[{x0}, {x1}) y:[{y0}, {y1})")
-            cfg["tiles"] = tiles
-            query["geopolygon"] = gs_bounds(self._gridspec, tiles)
+            cfg: Dict[str, Any] = dict(
+                grid=self._grid, freq=self._frequency,
+            )
 
-        if temporal_range is not None:
-            query.update(
-                temporal_range.dc_query(pad=0.6)
-            )  # pad a bit more than half a day on each side
-            cfg["temporal_range"] = temporal_range.short
+            if temporal_range is not None:
+                cfg["temporal_range"] = temporal_range.short
 
-        cfg["query"] = sanitize_query(query)
+            if tiles is not None:
+                cfg["tiles"] = tiles
 
         if DatasetCache.exists(self._output) and self._overwrite is False:
             raise ValueError(f"File database already exists: {self._output}")
 
-        msg("Connecting to the database, counting datasets")
-        n_dss = dataset_count(dc.index, **query)
-        if n_dss == 0:
-            msg("Found no datasets to process")
-            return False
-
         msg(f"Processing {n_dss:,d} datasets")
 
         msg("Training compression dictionary")
-        zdict = dictionary_from_product_list(dc, [product], samples_per_product=100)
+        dss, dss_copy = tee(dss)
```

Comment:
How smart is `tee`? my concern it will keep in RAM copy of unconsumed datasets from `dss_copy`.

probably safer to do islice to list, then copy into shuffle and train, then `chain` original list with the rest of the stream.

Reply:
the docs seem to agree with my hunch

> This itertool may require significant auxiliary storage (depending on how much temporary data needs to be stored). In general, if one iterator uses most or all of the data before another iterator starts, it is faster to use list() instead of tee().

Reply:
good catch! I had wrongly assumed copying an iterator was almost "free"